File: src_a_(C++)/demo1_10.c
// STRANNIK Modula-C-Pascal for Win32
// Demo program (Use Win32)
// Demo 2.10:Tennis (Classes)

include Win32

define hINSTANCE 0x400000
define className 'Strannik'

define RedColor 0x0000FF
define BlueColor 0xFF0000
define GreenColor 0x00FF00

define width 3

//------------------------ element ------------------------------

class element {
private:
  int x,y;
  int oldX,oldY;
  uint color;
public:
  void Initial(int ini_x, int ini_y, uint ini_color);
  int GetX();
  int GetY();
  int GetOldX();
  int GetOldY();
  int GetColor();
  void Move(int mov_x, int mov_y);
}

void element::Initial(int ini_x, int ini_y, uint ini_color)
{
  x=ini_x;
  y=ini_y;
  oldX=ini_x;
  oldY=ini_y;
  color=ini_color;
}

int element::GetX() {return x;}
int element::GetY() {return y;}
int element::GetOldX() {return oldX;}
int element::GetOldY() {return oldY;}
int element::GetColor() {return color;}

void element::Move(int mov_x, int mov_y)
{
  oldX=x;
  oldY=y;
  x+=mov_x;
  y+=mov_y;
}

//------------------------ region ------------------------------

class region:element {
private:
  int sizeX,sizeY;
public:
  void InitialRegion(int ini_x,int ini_y,int ini_sizeX,int ini_sizeY,uint ini_color);
  int GetSizeX()
  int GetSizeY()
  virtual void Paint(HDC dc, COLORREF back);
}

void region::InitialRegion(int ini_x,int ini_y,int ini_sizeX,int ini_sizeY,uint ini_color)
{
  this.Initial(ini_x,ini_y,ini_color);
  sizeX=ini_sizeX;
  sizeY=ini_sizeY;
}

int region::GetSizeX() {return sizeX;}
int region::GetSizeY() {return sizeY;}

void region::Paint(HDC dc, COLORREF back)
{
HPEN pen,oldpen; int colorBound,carX,carY; HBRUSH brush,oldbrush;

  if(back==0) {colorBound=this.GetColor(); carX=this.GetX(); carY=this.GetY();}
  else {colorBound=back; carX=this.GetOldX(); carY=this.GetOldY();}
  pen=CreatePen(PS_SOLID,width,colorBound); oldpen=SelectObject(dc,pen);
  brush=CreateSolidBrush(colorBound); oldbrush=SelectObject(dc,brush);
  Rectangle(dc,carX,carY,carX+sizeX,carY+sizeY);
  SelectObject(dc,oldpen); DeleteObject(pen);
  SelectObject(dc,oldbrush); DeleteObject(brush);
}

//------------------------ sphere ------------------------------

class sphere:element {
private:
  int diameter;
  int speedX,speedY;
public:
  void InitialSphere(int ini_x,int ini_y,int ini_speedX,int ini_speedY,int ini_diameter,uint ini_color);
  virtual void Paint(HDC dc, COLORREF back);
  bool MoveSphere(HWND wnd, region block,region screen);
}

void sphere::InitialSphere(int ini_x,int ini_y,int ini_speedX,int ini_speedY,int ini_diameter,uint ini_color)
{
  this.Initial(ini_x,ini_y,ini_color);
  diameter=ini_diameter;
  speedX=ini_speedX;
  speedY=ini_speedY;
}

void sphere::Paint(HDC dc, COLORREF back)
{
HPEN pen,oldpen; HBRUSH brush,oldbrush; int colorSphere,carX,carY;

  if(back==0) {colorSphere=this.GetColor(); carX=this.GetX(); carY=this.GetY();}
  else {colorSphere=back; carX=this.GetOldX(); carY=this.GetOldY();}
  pen=CreatePen(PS_SOLID,1,colorSphere); oldpen=SelectObject(dc,pen);
  brush=CreateSolidBrush(colorSphere); oldbrush=SelectObject(dc,brush);
  Ellipse(dc,carX-diameter/2,carY-diameter/2,carX+diameter/2,carY+diameter/2);
  SelectObject(dc,oldpen); DeleteObject(pen);
  SelectObject(dc,oldbrush); DeleteObject(brush);
}

bool sphere::MoveSphere(HWND wnd, region block,region screen)
{
int newX,newY;

  newX=this.GetX();
  newY=this.GetY();
  newX+=speedX; //move sphere
  newY+=speedY;
  if((newY>screen.GetSizeY())&&(  //test on contact with block
    (newX<block.GetX())||
    (newX>block.GetX()+block.GetSizeX())))
    return true;
  if(newX<0) {newX=-newX; speedX=-speedX;} //test on out of screen bound's
  if(newY<0) {newY=-newY; speedY=-speedY;}
  if(newX>screen.GetSizeX()) {newX=screen.GetSizeX()-(newX-screen.GetSizeX()); speedX=-speedX;}
  if(newY>screen.GetSizeY()) {newY=screen.GetSizeY()-(newY-screen.GetSizeY()); speedY=-speedY;}
  if(newX<0) {newX=-newX; speedX=-speedX;} //second test on out of screen bound's
  if(newY<0) {newY=-newY; speedY=-speedY;}
  if(newX>screen.GetSizeX()) {newX=screen.GetSizeX()-(newX-screen.GetSizeX()); speedX=-speedX;}
  if(newY>screen.GetSizeY()) {newY=screen.GetSizeY()-(newY-screen.GetSizeY()); speedY=-speedY;}
  this.Move(newX-this.GetX(),newY-this.GetY());
  return false;
}

//------------------------ main ------------------------------

  sphere ball;
  region block,screen;
  bool screenPaint;

bool wndProc(HWND wnd,uint msg,uint wparam,uint lparam)
{
HDC dc; PAINTSTRUCT stru; RECT reg;

  switch(msg) {
    case WM_CREATE:
      GetClientRect(wnd,reg);
      screen=new region; screen.InitialRegion(0,0,reg.right-1,reg.bottom-1,GreenColor);
      block=new region; block.InitialRegion(screen.GetSizeX()/2,screen.GetSizeY()-width,screen.GetSizeX()/10,width,RedColor);
      ball=new sphere; ball.InitialSphere(screen.GetSizeX()/2,1,5,25,10,BlueColor);
      screenPaint=false;
      SetTimer(wnd,1,100,NULL);
      return true;
      break;
    case WM_PAINT:
      dc=BeginPaint(wnd,stru);
      if(!screenPaint) {
        screen.Paint(dc,0);
        screenPaint=true;
      }
      ball.Paint(dc,screen.color);
      ball.Paint(dc,0);
      block.Paint(dc,screen.color);
      block.Paint(dc,0);
      EndPaint(wnd,stru);
      break;
    case WM_ERASEBKGND:return true; break;
    case WM_KEYDOWN:switch(loword(wparam)) {
      case VK_LEFT:
        if(block.GetX()>0) {
          block.Move(-block.GetSizeX()/2,0);
          InvalidateRect(wnd,NULL,true);
          UpdateWindow(wnd);
        }
        break;
      case VK_RIGHT:
        if(block.GetX()+block.GetSizeX()<screen.GetSizeX()) {
          block.Move(block.GetSizeX()/2,0);
          InvalidateRect(wnd,nil,true);
          UpdateWindow(wnd);
        }
        break;
      }
      break;
    case WM_TIMER:
      if(ball.MoveSphere(wnd,block,screen)) {
        KillTimer(wnd,1);
        MessageBox(0,"You lost.","Attention:",MB_ICONSTOP);
        DestroyWindow(wnd);
      }
      InvalidateRect(wnd,NULL,true);
      UpdateWindow(wnd);
      break;
    case WM_DESTROY:PostQuitMessage(0); return DefWindowProc(wnd,msg,wparam,lparam); break;
    default:return DefWindowProc(wnd,msg,wparam,lparam); break;
  }
}

  WNDCLASS cla;
  HWND wnd;
  MSG message;

void main()
{
//class registration
  with(cla) {
    style=CS_HREDRAW | CS_VREDRAW;
    lpfnWndProc=&wndProc;
    cbClsExtra=0;
    cbWndExtra=0;
    hInstance=hINSTANCE;    
    hIcon=0;
    hCursor=LoadCursor(0,pchar(IDC_ARROW));
    hbrBackground=COLOR_WINDOW;
    lpszMenuName=NULL;
    lpszClassName=className;
  }
  RegisterClass(cla);
                     
//create window
  wnd=CreateWindowEx(0,className,'Demo1_10',WS_OVERLAPPEDWINDOW,
    100,100,500,400, 0,0,hINSTANCE,NULL);

  ShowWindow(wnd,SW_SHOW);
  UpdateWindow(wnd);

//messages loop
  while(GetMessage(message,0,0,0)) {
    TranslateMessage(message);
    DispatchMessage(message);
  }

}

File: src_a_(Modula)/demo1_10.m
// STRANNIK Modula-C-Pascal for Win32
// Demo program
// Demo 10:Tennis (Classes)

module Demo1_10;
import Win32;

const 
  hINSTANCE=0x400000;
  className='Strannik';

  RedColor=0x0000FF;
  BlueColor=0xFF0000;
  GreenColor=0x00FF00;

  width=3;

//------------------------ element ------------------------------

type element=record ()
    x,y:integer;
    oldX,oldY:integer;
    color:cardinal;
    procedure Initial*(ini_x,ini_y:integer; ini_color:cardinal);
    procedure GetX*():integer;
    procedure GetY*():integer;
    procedure GetOldX*():integer;
    procedure GetOldY*():integer;
    procedure GetColor*():integer;
    procedure Move*(mov_x,mov_y:integer);
  end;

procedure (element) Initial(ini_x,ini_y:integer; ini_color:cardinal);
begin
  x:=ini_x;
  y:=ini_y;
  oldX:=ini_x;
  oldY:=ini_y;
  color:=ini_color;
end Initial;

procedure (element) GetX():integer; begin return x end GetX;
procedure (element) GetY():integer; begin return y end GetY;
procedure (element) GetOldX():integer; begin return oldX end GetOldX;
procedure (element) GetOldY():integer; begin return oldY end GetOldY;
procedure (element) GetColor():integer; begin return color end GetColor;

procedure (element) Move(mov_x,mov_y:integer);
begin
  oldX:=x;
  oldY:=y;
  inc(x,mov_x);
  inc(y,mov_y);
end Move;

//------------------------ region ------------------------------

type region=record (element)
    sizeX,sizeY:integer;
  end;

procedure (region) InitialRegion(ini_x,ini_y,ini_sizeX,ini_sizeY:integer; ini_color:cardinal);
begin
  self.Initial(ini_x,ini_y,ini_color);
  sizeX:=ini_sizeX;
  sizeY:=ini_sizeY;
end InitialRegion;

procedure (region) GetSizeX*():integer; begin return sizeX end GetSizeX;
procedure (region) GetSizeY*():integer; begin return sizeY end GetSizeY;

procedure (region) Paint(dc:HDC; back:COLORREF);
var pen,oldpen:HPEN; colorBound,carX,carY:integer; brush,oldbrush:HBRUSH;
begin
  if back=0
    then colorBound:=self.GetColor(); carX:=self.GetX(); carY:=self.GetY();
    else colorBound:=back; carX:=self.GetOldX(); carY:=self.GetOldY();
  end;
  pen:=CreatePen(PS_SOLID,width,colorBound); oldpen:=SelectObject(dc,pen);
  brush:=CreateSolidBrush(colorBound); oldbrush:=SelectObject(dc,brush);
  Rectangle(dc,carX,carY,carX+sizeX,carY+sizeY);
  SelectObject(dc,oldpen); DeleteObject(pen);
  SelectObject(dc,oldbrush); DeleteObject(brush);
end Paint;

//------------------------ sphere ------------------------------

type sphere=record (element)
    diameter:integer;
    speedX,speedY:integer;
  end;

procedure (sphere) InitialSphere(ini_x,ini_y,ini_speedX,ini_speedY,ini_diameter:integer; ini_color:cardinal);
begin
  self.Initial(ini_x,ini_y,ini_color);
  diameter:=ini_diameter;
  speedX:=ini_speedX;
  speedY:=ini_speedY;
end InitialSphere;

procedure (sphere) Paint(dc:HDC; back:COLORREF);
var pen,oldpen:HPEN; brush,oldbrush:HBRUSH; colorSphere,carX,carY:integer;
begin
  if back=0
    then colorSphere:=self.GetColor(); carX:=self.GetX(); carY:=self.GetY();
    else colorSphere:=back; carX:=self.GetOldX(); carY:=self.GetOldY();
  end;
  pen:=CreatePen(PS_SOLID,1,colorSphere); oldpen:=SelectObject(dc,pen);
  brush:=CreateSolidBrush(colorSphere); oldbrush:=SelectObject(dc,brush);
  Ellipse(dc,carX-diameter div 2,carY-diameter div 2,carX+diameter div 2,carY+diameter div 2);
  SelectObject(dc,oldpen); DeleteObject(pen);
  SelectObject(dc,oldbrush); DeleteObject(brush);
end Paint;

procedure (sphere) MoveSphere(wnd:HWND; block,screen:region):boolean;
var newX,newY:integer;
begin
  newX:=self.GetX();
  newY:=self.GetY();
  inc(newX,speedX); //moving sphere
  inc(newY,speedY);
  if (newY>screen.GetSizeY())and(  //test on contact with block
    (newX<block.GetX())or
    (newX>block.GetX()+block.GetSizeX())) then
    return true
  end;
  if newX<0 then newX:=-newX; speedX:=-speedX end; //test on out to screen bound's
  if newY<0 then newY:=-newY; speedY:=-speedY end;
  if newX>screen.GetSizeX() then newX:=screen.GetSizeX()-(newX-screen.GetSizeX()); speedX:=-speedX end;
  if newY>screen.GetSizeY() then newY:=screen.GetSizeY()-(newY-screen.GetSizeY()); speedY:=-speedY end;
  if newX<0 then newX:=-newX; speedX:=-speedX end; //second test on out to screen bound's
  if newY<0 then newY:=-newY; speedY:=-speedY end;
  if newX>screen.GetSizeX() then newX:=screen.GetSizeX()-(newX-screen.GetSizeX()); speedX:=-speedX end;
  if newY>screen.GetSizeY() then newY:=screen.GetSizeY()-(newY-screen.GetSizeY()); speedY:=-speedY end;
  self.Move(newX-self.GetX(),newY-self.GetY());
  return false
end MoveSphere;

//------------------------ main ------------------------------

var
  ball:sphere;
  block,screen:region;
  screenPaint:boolean;

procedure wndProc(wnd:HWND; msg,wparam,lparam:cardinal):boolean;
var dc:HDC; stru:PAINTSTRUCT; reg:RECT;
begin
  case msg of
    WM_CREATE:
      GetClientRect(wnd,reg);
      new(screen); screen.InitialRegion(0,0,reg.right-1,reg.bottom-1,GreenColor);
      new(block); block.InitialRegion(screen.GetSizeX() div 2,screen.GetSizeY()-width,screen.GetSizeX() div 10,width,RedColor);
      new(ball); ball.InitialSphere(screen.GetSizeX() div 2,1,5,25,10,BlueColor);
      screenPaint:=false;
      SetTimer(wnd,1,100,nil);
      return true;|
    WM_PAINT:
      dc:=BeginPaint(wnd,stru);
      if not screenPaint then
        screen.Paint(dc,0);
        screenPaint:=true
      end;
      ball.Paint(dc,screen.color);
      ball.Paint(dc,0);
      block.Paint(dc,screen.color);
      block.Paint(dc,0);
      EndPaint(wnd,stru);|
    WM_ERASEBKGND:return true;|
    WM_KEYDOWN:case loword(wparam) of
      VK_LEFT:
        if block.GetX()>0 then
          block.Move(-block.GetSizeX() div 2,0);
          InvalidateRect(wnd,nil,true);
          UpdateWindow(wnd)
        end;|
      VK_RIGHT:
        if block.GetX()+block.GetSizeX()<screen.GetSizeX() then
          block.Move(block.GetSizeX() div 2,0);
          InvalidateRect(wnd,nil,true);
          UpdateWindow(wnd)
        end;|
    end;|
    WM_TIMER:
      if ball.MoveSphere(wnd,block,screen) then
        KillTimer(wnd,1);
        MessageBox(0,"You lost.","Attention:",MB_ICONSTOP);
        DestroyWindow(wnd);
      end;
      InvalidateRect(wnd,nil,true);
      UpdateWindow(wnd);|
    WM_DESTROY:PostQuitMessage(0); return DefWindowProc(wnd,msg,wparam,lparam);|
    else return DefWindowProc(wnd,msg,wparam,lparam)
  end
end wndProc;

var
  cla:WNDCLASS;
  wnd:HWND;
  message:MSG;

begin

//class registration
  with cla do
    style:=CS_HREDRAW | CS_VREDRAW;
    lpfnWndProc:=addr(wndProc);
    cbClsExtra:=0;
    cbWndExtra:=0;
    hInstance:=hINSTANCE;    
    hIcon:=0;
    hCursor:=LoadCursor(0,pstr(IDC_ARROW));
    hbrBackground:=COLOR_WINDOW;
    lpszMenuName:=nil;
    lpszClassName:=className;
  end;
  RegisterClass(cla);
                     
//window create
  wnd:=CreateWindowEx(0,className,'Demo1_10',WS_OVERLAPPEDWINDOW,
    100,100,500,400, 0,0,hINSTANCE,nil);

  ShowWindow(wnd,SW_SHOW);
  UpdateWindow(wnd);

//messages loop
  while GetMessage(message,0,0,0) do
    TranslateMessage(message);
    DispatchMessage(message);
  end;

end Demo1_10.

